### ISSUE TITLE
electron wrapper

### ISSUE THREAD

**Issue by davidmarkclements** [davidmarkclements/0x#30]
display flamegraph.html in electron

1. should this be default behaviour (of a v2)
2. should it be behind a flag
3. if behind a flag, should we make users download electron etc. during default install if they don't intend to use this feature (e.g. running 0x on server side)

**Comment by mcdonnelldean** [comment]
If not picked up by Monday let me know I can get someone on it

**Comment by davidmarkclements** [comment]
@eljefedelrodeodeljefe can have first dibs on this - if he wants it - since he's made a lot of great contributions already - but great to know help is there - no harm if we split the task into pieces either

on that here's how I think we should do this, any input appreciated

1. have the ability to automatically "open with" any specified command, e.g. the `open` command on OSX  so it automatically opens in the default browser
2. use this same ability to open with electron, possibly overlaw a convenience flag for this special case
3. when opening with electron, we mean opening with the electron app that enhances the flame graph interaction experience - primarily by allowing file system access via the stacks - so possibly this app that enhances the experience should be a separate tool, that's separately installed

@mcollina @mafintosh @eljefedelrodeodeljefe @mcdonnelldean - what do you think?

**Comment by mafintosh** [comment]
although the "open with" idea sounds nice and simple a "pure" electron integration would mean that it wouldn't need to write any files to disk (assuming that is even possible to configure) which is nice if you a bunch of profiles while testing stuff out

**Comment by mafintosh** [comment]
so :+1: on 3.

**Comment by mcdonnelldean** [comment]
I'm good with all that 👌 Oooh can't wait to play with this when done! Yaaay

**Comment by mafintosh** [comment]
Crazy idea: it could even have a mode where it re-ran benchmarks and updated the flamegraph when underlying deps changed. Also side-by-side comparison and stuff with older versions.

**Comment by davidmarkclements** [comment]
@mafintosh good point so we've got two scenarios for an electron app here

1. instant viewer, no trace of files on system - put them in /tmp if we need them - just make the experience seamless
2. viewer of profiles that we have kept, e.g. cases where we want the profile folder, or cases where we want to view a profile generated some time in the past 

it's interesting you should say that, because @mcollina and I were discussing this very thing, along with branch comparisons etc., and integrating http://npm.im/autocannon for http benchmarks etc. - we we're calling it the Performance Toolkit.

This could be the start of that, which is another reason I'm suggesting as a separate app that integrates

**Comment by davidmarkclements** [comment]
so here's the current order of events:

```
node (0x) -> perf/dtrace -> node (profile process) -> chrome (flamegraph.html)
```

We have two possibilities here:

The simple integration:

```
node (0x) -> perf/dtrace -> node (profile process) -> electron (flamegraph.html)
```

The tighter integration:

```
electron (0x) -> perf/dtrace -> node (profile process) -> electron (0x)
```

Benefits of the tighter integration *could* be generating everything in memory, not sure if this is worth it as opposed to simply dumping files in /tmp - also large profiles can blow the stack currently (https://github.com/davidmarkclements/0x/issues/27)

**Comment by yoshuawuyts** [comment]
Not that I'm assuming you would, but if `electron` integration is added, could it not be added as a hard dependency to this package?

With electron being ~`120mb`, it would tremendously inflate the amount of files downloaded. I'm currently adding `0x` to a good number of my projects, and it would be a show stopper on a 110GB HDD (300 existing projects * 0.12GB = 36GB of electron installs).

I know things like `pnpm` can partially mitigate this, but for I'd much rather have `0x` exist as-is with `electron` being optional / living somewhere else. Thanks! ☺️

**Comment by mcollina** [comment]
I agree with @yoshuawuyts. I would rather prefer the electron-based 0x to be a standard desktop app, maybe even with a command line launcher, but something you download/install externally rather than something that sits in the NPM tree. Generally I do think distributing an electron app via NPM is not the proper way to do it. We can still do it, but as an interim solution.

However, defaulting to open the file on the default browser will be highly welcomed. a `--no-open` flag could always be passed through.

**Comment by davidmarkclements** [comment]
but on a server environment you want no-open as default, also with open you want to specify what it's opening with - this means we need a user settings config file

actions (once agreed, some of these should become issues)

1) user settings config file (ala rc?)
2) `open` flag / setting w/ corresponding no-open when open is default
3) entirely new 0x-viewer app
4) support 0x-viewer with `open` flag / setting (just `npm i -g 0x-viewer`)

**Comment by davidmarkclements** [comment]
@yoshuawuyts just curious - how come you're adding 0x to individual projects as opposed to using the tool globally?

**Comment by mcollina** [comment]
@davidmarkclements alway running `'open'` should cause no harm on a server with no X daemon running.

**Comment by yoshuawuyts** [comment]
@davidmarkclements I like my packages to be self-contained. Same reason why I don't rely on globally installed versions of `standard` for example.

**Comment by davidmarkclements** [comment]
@mcollina true - but I'm still not sure about defaulting to a command for opening (since it's system dependent) - I'd rather have it configurable

@yoshuawuyts  - me too, I guess it's about where you draw the line though, for me introspection tools are system wide - e.g. I wouldn't include node-inspector with every project. But at the same time, I see your point - there's a degree of automation you can achieve with 0x that can make it useful as part of a build pipeline - then including it makes sense

**Comment by eljefedelrodeodeljefe** [comment]
Has someone started already? Would do it exactly now and catch up to the discussion. Sorry for the delay.

**Comment by mcollina** [comment]
@davidmarkclements just saying: https://www.npmjs.com/package/open

**Comment by mafintosh** [comment]
@eljefedelrodeodeljefe dont think so. go build something awesome and avoid the bike shed :tada:
